cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp.hpp

```diff
@@ -0,0 +1,459 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/backend/common.hpp"
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+#include "oneapi/dal/backend/primitives/distance.hpp"
+#include "oneapi/dal/backend/primitives/sort/sort.hpp"
+
+namespace oneapi::dal::kmeans::backend {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+namespace bk = dal::backend;
+namespace pr = dal::backend::primitives;
+
+inline std::int64_t get_recommended_sg_size(sycl::queue& queue) {
+    // TODO optimization/dispatching
+    return 16;
+}
+
+inline std::int64_t get_recommended_wg_count(sycl::queue& queue) {
+    // TODO optimization/dispatching
+    return 128;
+}
+
+inline std::int64_t get_scaled_wg_size_per_row(const sycl::queue& queue,
+                                               std::int64_t column_count,
+                                               std::int64_t preffered_wg_size) {
+    const std::int64_t sg_max_size = bk::device_max_sg_size(queue);
+    const std::int64_t row_adjusted_sg_num =
+        column_count / sg_max_size + std::int64_t(column_count % sg_max_size > 0);
+    std::int64_t expected_sg_num = std::min(preffered_wg_size / sg_max_size, row_adjusted_sg_num);
+    if (expected_sg_num < 1)
+        expected_sg_num = 1;
+    return dal::detail::check_mul_overflow(expected_sg_num, sg_max_size);
+}
+
+inline std::int64_t get_gpu_sg_size(sycl::queue& queue) {
+    // TODO optimization/dispatching
+    return 16;
+}
+
+namespace bk = dal::backend;
+namespace pr = dal::backend::primitives;
+
+template <typename T>
+struct select_min_distance {};
+
+template <typename T>
+struct set_indices {};
+
+template <typename T>
+struct find_candidates_kernel {};
+
+template <typename T>
+struct fill_empty_cluster_kernel {};
+
+template <typename T>
+struct centroid_reduction {};
+
+template <typename T>
+struct reset_partial_centroids {};
+
+template <typename T>
+struct centroid_merge {};
+
+template <typename T>
+struct compute_obj_function {};
```

Comment:
Is this some sort of workaround?

Reply:
Yes. It is a list of names for all lambdas. There is an idea that it helps with "buildProgram" issue but I am not 100% sure. Nevertheless, it would be good to remain it until the issue is resolved by compiler.